akka-docs/rst/scala/code/docs/persistence/query/PersistenceQueryDocSpec.scala

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package docs.persistence.query
+
+import akka.actor._
+import akka.persistence.query.scaladsl.ReadJournal
+import akka.persistence.{ Recovery, PersistentActor }
+import akka.persistence.query._
+import akka.stream.{ FlowShape, ActorMaterializer }
+import akka.stream.scaladsl.FlowGraph
+import akka.stream.scaladsl.{ Flow, Sink, Source }
+import akka.testkit.AkkaSpec
+import akka.util.Timeout
+import docs.persistence.query.PersistenceQueryDocSpec.TheOneWhoWritesToQueryJournal
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+
+object PersistenceQueryDocSpec {
+
+  implicit val timeout = Timeout(3.seconds)
+
+  class NoopReadJournal(sys: ExtendedActorSystem) extends ReadJournal {
+    override def query[T, M](q: Query[T, M], hints: Hint*): Source[T, M] =
+      Source.empty.mapMaterializedValue(_ => null.asInstanceOf[M])
+  }
+
+  //#my-read-journal
+  class MyReadJournal(system: ExtendedActorSystem) extends ReadJournal {
+
+    // TODO from config
+    private val defaulRefreshInterval: FiniteDuration = 3.seconds
+
+    override def query[T, M](q: Query[T, M], hints: Hint*): Source[T, M] =
+      q match {
+        case EventsByTag(tag, offset) ⇒
+          val props = MyEventsByTagPublisher.props(tag, offset, refreshInterval(hints))
+          Source.actorPublisher[EventEnvelope](props)
+            .mapMaterializedValue(_ ⇒ noMaterializedValue)
```

Comment:
> It might be that the type signature is too ambitious …

True, on one hand - from the user's point of view it's great; the implementations though need to perform casts. We would really like to make use of types "based on the query": we know that *"query for persistence ids"* == `Source[String, _]`, so it's `Query[String, Unit]`. For the user of the API it makes complete sense; I did not find a way we could preserve typesafety for both ends, something to investigate further I guess?